### DISCUSSION TITLE
Mill doesn't watch or restarts zio-http projects

### DISCUSSION THREAD

**Discussion by carlosedp** [com-lihaoyi/mill#2077] [Q&A] [answered]
I've set a sample zio-http project similar to https://github.com/zio/zio-http#getting-started.

When running `mill -w sample.run`, mill doesn't watch the sources and neither restarts when files change. Watching the `test` target works fine tho.

On zio-http page, it mentions using sbt-revolver plugin for watching sources during dev.

**Comment by lefou** [comment, marked as answer]
IIRC, sbt-revolver is interrupting a running process when it detects changes. Mills `JavaModule.run` does not. Instead, it would re-run once the process has ended and only after  changes were detected.

If you want your server process to stop and restart whenever its sources have changed, you can try to use `JavaModule.runBackground`. 

```bash
$ mill inspect main.runBackground
[1/1] inspect 
main.runBackground(JavaModule.scala:697)
    Runs this module's code in a background process, until it dies or
    `runBackground` is used again. This lets you continue using Mill while
    the process is running in the background: editing files, compiling, and
    only re-starting the background process when you're ready.
    
    You can also use `-w foo.runBackground` to make Mill watch for changes
    and automatically recompile your code & restart the background process
    when ready. This is useful when working on long-running server processes
    that would otherwise run forever

Inputs:
    main.runClasspath
    main.forkArgs
    main.forkEnv
    main.finalMainClass
    main.forkWorkingDir
    main.runUseArgsFile
```

To stop such background process, you can run `mill clean sample.runBackground`.

If this is not what you're looking for and you still think you discovered an issue, please provide a [MCVE](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by carlosedp** [comment]
Awesome, just tested with `mill -w backend.runBackground` and it reloaded fine on changes!
Thanks Tobias!

**Comment by carlosedp** [comment]
BTW, if I run `mill -w backend.runBackground` it's not really in "background" since it's watching the sources and the prompt doesnt return. Also if I Ctrl-C, it exits and doesn't need the `mill clean backend.runBackground` command.

**Reply by ajrnz** [reply]
It only works like that in when run with `-w` - technically the process is still started in the background.

Be a bit careful as it's slightly flakey. About 1 in 20 times (on my macs) it fails to shutdown the task on Ctrl-C

**Comment by carlosedp** [comment]
Noticed that as well, if you do:

`bash -c "trap 'mill clean backend.runBackground' EXIT; mill -w backend.runBackground"` it will trap the exit signal and run the cleanup.
